<a href="https://colab.research.google.com/github/kimsam0103/demo_test/blob/master/Module%200/Session%206/m0s6nb05_nested_data_NBA_exercises_final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Nested Data Structures: NBA Player Analysis -- Part 2

## This notebook is an application of the NBA nested data notebook.

## We will be taking this nested data structure and working through what might be a typical 2 or 3 point exam exercise. This will require us to execute multiple steps to come to our final answer.

### We will use a helper function as a part of the solution.

In [1]:
!wget https://raw.githubusercontent.com/gt-cse-6040/bootcamp/main/Module%200/Session%206/nba_json.txt

--2024-01-30 00:19:08--  https://raw.githubusercontent.com/gt-cse-6040/bootcamp/main/Module%200/Session%206/nba_json.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 8903 (8.7K) [text/plain]
Saving to: ‘nba_json.txt’

nba_json.txt        100%[===================>]   8.69K  --.-KB/s    in 0s      

2024-01-30 00:19:08 (69.3 MB/s) - ‘nba_json.txt’ saved [8903/8903]



#### Data loading. This will normally be provided in the exam, and the students will not be required to load data from input files.

In [2]:
# loading the data, same as the previous NB
# import the required package
import ast

# reading the data from the json file
with open('nba_json.txt') as f:
    data = f.readlines()

nested = [] # this is going to be our nested data structure
#put each line(string) into a dictionary
for line in data:
    result = ast.literal_eval(line)
#     print(result)
    nested.append(result)

### As we know from the previous NB, the key 'resultSets' contains the data the we will want to manipulate.

### Let's take a look at it again. This step will generally be provided on the exam, for the student to be able to visualize the source data that is to be analyzed.

In [7]:
print(nested[0]['resultSets'][0])
print('\n')

{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 355.6], ['22019', 201567, 'Kevin Love', 202329, 'Al-Farouq Aminu', 2, '5:17', 25.0, 4, 14, 0, 1, 0, 2, 3, 0.667, 0, 1, 

In [15]:
type(nested)

for i in range(len(nested)):
  print(type(nested[i]))

<class 'dict'>
<class 'dict'>
<class 'dict'>
<class 'dict'>
<class 'dict'>
<class 'dict'>
<class 'dict'>


In [16]:
print(nested[6]['resultSets'][0]['rowSet'][0][0:5])
print('\n')

['22019', 1628374, 'Lauri Markkanen', 203507, 'Giannis Antetokounmpo']




In [9]:
print(nested[0]['resultSets'][0]['headers'][0:5])
print('\n')

['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME']




In [10]:
print(nested[0]['resultSets'][0]['rowSet'][0][0:5])
print('\n')

['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu']




In [3]:
print(nested[0]['resultSets'])
print('\n')
print(nested[0]['resultSets'][0].keys())

[{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 355.6], ['22019', 201567, 'Kevin Love', 202329, 'Al-Farouq Aminu', 2, '5:17', 25.0, 4, 14, 0, 1, 0, 2, 3, 0.667, 0, 1,

## Note that this is a list containing a single dictionary, with three keys. Remember also that we showed this in Python Tutor, in the previous notebook.

#### The key 'rowSet' contains the statistics that we are interested in, and the key 'headers' defines what each of the statisics is.

#### Note that this first dict is for the player Al-Farouq Aminu, with the player_id 202329.

In [4]:
print(nested[0]['resultSets'][0]['headers'])
print('\n')
print(nested[0]['resultSets'][0]['rowSet'])
print('\n')
print(nested[0]['resultSets'][0]['rowSet'][0])
# print('\n')
# print(nested[0]['resultSets'][0]['rowSet'][1])

['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC']


[['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 355.6], ['22019', 201567, 'Kevin Love', 202329, 'Al-Farouq Aminu', 2, '5:17', 25.0, 4, 14, 0, 1, 0, 2, 3, 0.667, 0, 1, 0.0, 0, 0, 0, 0, 0, 0, 0, 316.8]]


['22019', 1

## Exercise 0 (3 points)

### Who are the players that each defensive player guarded during the season, and how effective was the defensive player against each?

### Write a function, `players_guarded_minutes_dict_sorted(var_stats, DEF_PLAYER_NAME)` that outputs a list of dictionaries.

1. Within the list, there will be an individual dictionary for each defensive player name.

2. The dictionary key is the defensive player's name, and the value is a list of tuples.

    a. The first item of each tuple is the name of the offensive players he guarded, and the second item of the tuple is the number of points allowed.

    b. The list of tuples should be sorted by the number of points allowed, from most to least.

    c. If the points are the same, then sort by the name in alphabetical order.
    
#### Note that items b and c, the sorting, are requirements that are VERY COMMON to see on exams.

#### For the function, the `var_stats` parameter will be the `nested` variable. This is a list of dictionaries.

#### The `DEF_PLAYER_NAME` parameter will be a list of the defensive players to include in the output dictionary.

For example, an input might be:

In [11]:
var_stats = (nested)
DEF_PLAYER_NAME = ['Al-Farouq Aminu','Bam Adebayo','Giannis Antetokounmpo',
                   'Nickeil Alexander-Walker', 'Kyle Anderson', 'Jarrett Allen', 'Grayson Allen']
# print(var_stats)
# print(DEF_PLAYER_NAME)

#### To solve the problem, your function might do the following:

1. Create an empty list to return (overall list)

2. Loop over the list of the defensive players

    a. For each defensive player, find the nested dictionary with this player's information. The dictionary will be at the following level:

        `players_guarded_dict(nested[#]['resultSets'][0]`.


    b. Loop over the players guarded -- You might want to use a helper function for the below activities
        i. create each tuple, with name and points allowed.

        ii. Append each tuple to the list for the player

        iii. Sort the list of tuples

    e. Populate the dictionary for that player, with key being the defensive player name and the value being the list of tuples.
    
    f. Append the overall list with that player's dictionary
    
3. Return the overall list

## <span style="background-color: #FFFF00"> If you see something like this on an exam, this is giving you the steps that you will want to take for your solution.</span>

## <span style="background-color: #FFFF00"> So these are the steps of your `STRATEGY` in the `Problem Solving Method`.</span>

#### If your code executes successfully, the output will be:

In [17]:
final_result = [{'Al-Farouq Aminu': [('OG Anunoby', 5), ('Kevin Love', 4), ('Rui Hachimura', 4), ('Trey Lyles', 3)]},
                {'Bam Adebayo': [('Joel Embiid', 31), ('Nikola Vucevic', 22), ('Myles Turner', 13), ('John Collins', 10)]},
                {'Giannis Antetokounmpo': [('Kevin Love', 23), ('Lauri Markkanen', 11), ('Ben Simmons', 5), ('Myles Turner', 5)]},
                {'Nickeil Alexander-Walker': [('Jordan Clarkson', 16), ('Seth Curry', 5), ('Jerome Robinson', 0)]},
                {'Kyle Anderson': [('Joe Ingles', 21), ('Harrison Barnes', 17), ('Aaron Gordon', 6), ('Nemanja Bjelica', 3)]},
                {'Jarrett Allen': [('Bam Adebayo', 27), ('Andre Drummond', 21), ('Nikola Vucevic', 21), ('Daniel Theis', 13)]},
                {'Grayson Allen': [('Donovan Mitchell', 11), ('Bryn Forbes', 4), ('Kyle Korver', 3)]}]

#### How are we going to find the values that we need?

In [18]:
var_stats = (nested)
DEF_PLAYER_NAME = ['Al-Farouq Aminu','Bam Adebayo','Giannis Antetokounmpo',
                   'Nickeil Alexander-Walker', 'Kyle Anderson', 'Jarrett Allen', 'Grayson Allen']

In [80]:
# for i in DEF_PLAYER_NAME:
#   print(i)
# aa = nested[0]['resultSets'][0]['headers']
# bb = nested[0]['resultSets'][0]['rowSet'][0]
# print(aa)
# print(bb)

pts = nested[0]['resultSets'][0]['rowSet'][0][8]
def_player_name = nested[0]['resultSets'][0]['rowSet'][0][4]
off_player_name = nested[0]['resultSets'][0]['rowSet'][0][2]
dict_def={def_player_name: (off_player_name, pts)}

print(dict_def)



{'Al-Farouq Aminu': ('Rui Hachimura', 4)}


In [51]:
print(len(nested))

7


[{'Al-Farouq Aminu': [('OG Anunoby', 5),
   ('Kevin Love', 4),
   ('Rui Hachimura', 4),
   ('Trey Lyles', 3)]},
 {'Bam Adebayo': [('Joel Embiid', 31),
   ('Nikola Vucevic', 22),
   ('Myles Turner', 13),
   ('John Collins', 10)]},
 {'Giannis Antetokounmpo': [('Kevin Love', 23),
   ('Lauri Markkanen', 11),
   ('Ben Simmons', 5),
   ('Myles Turner', 5)]},
 {'Nickeil Alexander-Walker': [('Jordan Clarkson', 16),
   ('Seth Curry', 5),
   ('Jerome Robinson', 0)]},
 {'Kyle Anderson': [('Joe Ingles', 21),
   ('Harrison Barnes', 17),
   ('Aaron Gordon', 6),
   ('Nemanja Bjelica', 3)]},
 {'Jarrett Allen': [('Bam Adebayo', 27),
   ('Andre Drummond', 21),
   ('Nikola Vucevic', 21),
   ('Daniel Theis', 13)]},
 {'Grayson Allen': [('Donovan Mitchell', 11),
   ('Bryn Forbes', 4),
   ('Kyle Korver', 3)]}]

In [86]:
nested_list = nested[0]['resultSets'][0]['rowSet']
type(nested_list)
print(len(nested_list))

final_list=[]

for j in range(len(nested)):
  def_player_name = nested[j]['resultSets'][0]['rowSet'][0][4]
  # print(j, " | ", DEF_PLAYER_NAME[j], " | ", def_player_name)


  off_pts_list = []

  for i in range(len(nested[j]['resultSets'][0]['rowSet'])):
    # print(i," | ", nested[j]['resultSets'][0]['rowSet'])
    pts = nested[j]['resultSets'][0]['rowSet'][i][8]
    def_player_name = nested[j]['resultSets'][0]['rowSet'][i][4]
    off_player_name = nested[j]['resultSets'][0]['rowSet'][i][2]
    off_pts_list.append((off_player_name, pts))

  sorted_off_pts_list = sorted(off_pts_list, key = lambda x: (-x[1], x[0]))
  # sorted_off_pts_list = sorted(off_pts_list, key = lambda x: x[1], reverse=True)


  def_dict = {def_player_name: sorted_off_pts_list}

  final_list.append(def_dict)

display(sorted_final_list)

4


[{'Al-Farouq Aminu': [('OG Anunoby', 5),
   ('Kevin Love', 4),
   ('Rui Hachimura', 4),
   ('Trey Lyles', 3)]},
 {'Bam Adebayo': [('Joel Embiid', 31),
   ('Nikola Vucevic', 22),
   ('Myles Turner', 13),
   ('John Collins', 10)]},
 {'Giannis Antetokounmpo': [('Kevin Love', 23),
   ('Lauri Markkanen', 11),
   ('Ben Simmons', 5),
   ('Myles Turner', 5)]},
 {'Grayson Allen': [('Donovan Mitchell', 11),
   ('Bryn Forbes', 4),
   ('Kyle Korver', 3)]},
 {'Jarrett Allen': [('Bam Adebayo', 27),
   ('Andre Drummond', 21),
   ('Nikola Vucevic', 21),
   ('Daniel Theis', 13)]},
 {'Kyle Anderson': [('Joe Ingles', 21),
   ('Harrison Barnes', 17),
   ('Aaron Gordon', 6),
   ('Nemanja Bjelica', 3)]},
 {'Nickeil Alexander-Walker': [('Jordan Clarkson', 16),
   ('Seth Curry', 5),
   ('Jerome Robinson', 0)]}]

In [7]:
print('What does each dictionary look like, in var_stats?')
print(var_stats[0]['resultSets'][0])
print('\n')

print('List of statistics for each player matchup')
print(var_stats[0]['resultSets'][0]['rowSet'][0])
print('\n')

print('defensive player name')
print(var_stats[0]['resultSets'][0]['rowSet'][0][4])
print('\n')

print('offensive player name')
print(var_stats[0]['resultSets'][0]['rowSet'][0][2])
print('\n')

print('points scored/allowed')
print(var_stats[0]['resultSets'][0]['rowSet'][0][8])

What does each dictionary look like, in var_stats?
{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 355.6], ['22019', 201567, 'Kevin Love', 202329, 'Al-Farouq Aminu', 2

In [8]:
def players_guarded_minutes_dict_sorted(var_stats, DEF_PLAYER_NAME):
    ###YOUR CODE HERE

#     helper function
    def create_tuples_list(stats):
        player_list = []
        for player_guarded in stats:
            player_tuple = (player_guarded[2],player_guarded[8])
#                     print(player_tuple)
            player_list.append(player_tuple)

#         this code uses a lambda function, which we have not yet covered
#         don't worry about understanding it yet
#         but you will want to have this code snippet on your "speed dial"
#         of useful code examples for the exam

        # https://www.geeksforgeeks.org/python-mutiple-column-sort-in-tuples/
        # https://www.geeksforgeeks.org/python-program-to-sort-a-list-of-tuples-by-second-item/
        player_list = sorted(player_list, key = lambda x: (-x[1], x[0]))
        return player_list


    ret_list = []


    for player in DEF_PLAYER_NAME:

        for item in range(0,len(var_stats)):

            player_dict = {}

#             get the list of players guarded stats
            stats = (var_stats[item]['resultSets'][0]['rowSet'])

            if (stats[0][4] == player):
#                 call the helper function
                tup_list = create_tuples_list(stats)

                player_dict[player] = tup_list

                ret_list.append(player_dict)

    return ret_list

# use for testing
# print(players_guarded_minutes_dict_sorted(var_stats, DEF_PLAYER_NAME))

In [ ]:
final_list=[]

for name in DEF_PLAYER_NAME:
  for j in range(len(nested)):
    def_player_name = nested[j]['resultSets'][0]['rowSet'][0][4]

    if name == def_player_name:
      off_pts_list = []

      for i in range(len(nested[j]['resultSets'][0]['rowSet'])):
        # print(i," | ", nested[j]['resultSets'][0]['rowSet'])
        pts = nested[j]['resultSets'][0]['rowSet'][i][8]
        def_player_name = nested[j]['resultSets'][0]['rowSet'][i][4]
        off_player_name = nested[j]['resultSets'][0]['rowSet'][i][2]
        off_pts_list.append((off_player_name, pts))

      sorted_off_pts_list = sorted(off_pts_list, key = lambda x: (-x[1], x[0]))
      # sorted_off_pts_list = sorted(off_pts_list, key = lambda x: x[1], reverse=True)


      def_dict = {name: sorted_off_pts_list}

      final_list.append(def_dict)

display(final_list)

In [96]:
def players_guarded_minutes_dict_sorted1(var_stats, DEF_PLAYER_NAME):
  nested = var_stats
  final_list=[]
  for name in DEF_PLAYER_NAME:
    for j in range(len(nested)):
      def_player_name = nested[j]['resultSets'][0]['rowSet'][0][4]

      if name == def_player_name:
        off_pts_list = []

        for i in range(len(nested[j]['resultSets'][0]['rowSet'])):
          # print(i," | ", nested[j]['resultSets'][0]['rowSet'])
          pts = nested[j]['resultSets'][0]['rowSet'][i][8]
          def_player_name = nested[j]['resultSets'][0]['rowSet'][i][4]
          off_player_name = nested[j]['resultSets'][0]['rowSet'][i][2]
          off_pts_list.append((off_player_name, pts))

        sorted_off_pts_list = sorted(off_pts_list, key = lambda x: (-x[1], x[0]))
        # sorted_off_pts_list = sorted(off_pts_list, key = lambda x: x[1], reverse=True)


        def_dict = {name: sorted_off_pts_list}

        final_list.append(def_dict)

  return final_list

In [97]:
# TEST CELL
# test code
OUTPUT_LIST = players_guarded_minutes_dict_sorted1(var_stats, DEF_PLAYER_NAME)
display(OUTPUT_LIST)

[{'Al-Farouq Aminu': [('OG Anunoby', 5),
   ('Kevin Love', 4),
   ('Rui Hachimura', 4),
   ('Trey Lyles', 3)]},
 {'Bam Adebayo': [('Joel Embiid', 31),
   ('Nikola Vucevic', 22),
   ('Myles Turner', 13),
   ('John Collins', 10)]},
 {'Giannis Antetokounmpo': [('Kevin Love', 23),
   ('Lauri Markkanen', 11),
   ('Ben Simmons', 5),
   ('Myles Turner', 5)]},
 {'Nickeil Alexander-Walker': [('Jordan Clarkson', 16),
   ('Seth Curry', 5),
   ('Jerome Robinson', 0)]},
 {'Kyle Anderson': [('Joe Ingles', 21),
   ('Harrison Barnes', 17),
   ('Aaron Gordon', 6),
   ('Nemanja Bjelica', 3)]},
 {'Jarrett Allen': [('Bam Adebayo', 27),
   ('Andre Drummond', 21),
   ('Nikola Vucevic', 21),
   ('Daniel Theis', 13)]},
 {'Grayson Allen': [('Donovan Mitchell', 11),
   ('Bryn Forbes', 4),
   ('Kyle Korver', 3)]}]

In [98]:
assert(OUTPUT_LIST == final_result), 'Your output does not match the solution'
print('passed')

passed
